keps/sig-auth/4317-pod-certificates/README.md

```diff
@@ -0,0 +1,1274 @@
+<<[/UNRESOLVED]>>
+
+#### Why a New Type?
+
+The following constraints led to the adoption of a new API type, as opposed to extending CertificateSigningRequest:
+1. We wanted the pod details to be legible without having to parse an ASN.1 CSR.
+2. We wanted the pod details (especially the associated node name) to be
+  available for use in an informer selector.  While these could be surfaced as labels on CertificateSigningRequest objects, labels are immutable.
+3. We wanted pod details to be immutable.
+4. We wanted to avoid putting Pod Certificate-specific fields in the generic
+  CertificateSigningRequest.
+5. Avoid confusing existing signer implementations with pod information they
+   don't understand.
+
+The available options were:
+
+* Label CertificateSigningRequests with node/pod labels by convention.  This didn't meet goal 3.
+* Hoist pod fields into CSR spec.  This didn't meet goal 4.
+* Embed pod details into CSR spec.request field.  This didn't meet goals 1 and 2.
+* New dedicated PodCertificateRequest type. Meets all goals.
+
+#### API Definition
+
+```
+// PodCertificateRequest encodes a pod requesting a certificate from a given
+// signer.
+//
+// Kubelets use this API to implement podCertificate projected volumes
+type PodCertificateRequest struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// spec contains the details about the certificate being requested.
+	Spec PodCertificateRequestSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+
+	// status contains the issued certificate, and a standard set of conditions.
+	// +optional
+	Status PodCertificateRequestStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// PodCertificateRequestSpec describes the certificate request.  All fields are
+// immutable after creation.
+type PodCertificateRequestSpec struct {
+	// signerName indicates the request signer.
+	SignerName string `json:"signerName" protobuf:"bytes,1,opt,name=signerName"`
+
+	// podName is the name of the pod into which the certificate will be mounted.
+	PodName string `json:"podName" protobuf:"bytes,2,opt,name=podName"`
+	// podUID is the UID of the pod into which the certificate will be mounted.
+	PodUID types.UID `json:"podUID" protobuf:"bytes,3,opt,name=podUID"`
+
+	// serviceAccountName is the name of the service account the pod is running as.
+	ServiceAccountName string `json:"serviceAccountName" protobuf:"bytes,4,opt,name=serviceAccountName"`
+
+	// nodeName is the name of the node the pod is assigned to.
+	NodeName string `json:"nodeName" protobuf:"bytes,6,opt,name=nodeName"`
+
+	// pkixPublicKey is the PKIX-serialized public key the signer should issue
+	// the certificate to.
+	//
+	// The key must be one of RSA-3072, RSA-4096, ECDSA-P256 or ECDSA-P384.
```

Comment:
clarify this list of key types could be expanded in the future, and what a signer should do if it encountered a key type it does not support

Reply:
Done.